### ISSUE TITLE
Add `Either#chainLeft`

### ISSUE THREAD

**Issue by EricCrosson** [gigobyte/purify#135]
Hi there, 

great project! I hardly go a day without using purify-ts. A thousand thanks for your work to date!

I'm opening this issue as a place to discuss the addition of a `chainLeft` function for the `Either` type, which is like `chain` but for the rejection branch. I'd define the function as follows

```
Either a b ~> (a -> Either c b) -> Either c b
```

so that `Right` values are kept the same and `Left` values are chained on the provided function. While `chainLeft` doesn't appear in the fantasy-land spec, I've found a function like this helpful in "resetting" a computation back on to the successful path, in certain situations from which an error can be recovered (but without losing the surrounding type as with `orDefault`).

This signature represents one take, I guess after thinking about it there could be other signatures, such as

```
Either a b ~> (a -> Either c d) -> Either c b|d
```

in the event that the function provided to `chainLeft` returns a different type than `b`.

Nothing concrete for now obviously, just wanted to touch base. 

Cheers

**Comment by EricCrosson** [comment]
Here's an example of a fantasy-land-compliant project offering `chainRej`:

https://github.com/fluture-js/Fluture/tree/11.x#chainrej

**Comment by gigobyte** [comment]
If you can recover from an error why would you want to keep the value wrapped in an Either? I think that if you can `chainLeft` and return a Right you might as well just use `orDefault` like you mentioned since you know that the value is always Right.

**Comment by EricCrosson** [comment]
That's a fair question. The best answer I have is to enable further transformation on the wrapped value, as in the case where recovery happened in the middle of a chain of transformations but more transformations still have to be applied, and some of those may fail (irrecoverably). Without `chainLeft`, further chaining necessitates using `orDefault` to drop out of the `Either` context and then re-wrapping, which is less readable and writable in my opinion.

```
function getConfigSubsection() {

    const config = readConfig()
        .toEither(`Could not read config file`)
        .chain(ConfigFile.decode)
        .mapLeft(error => (console.error(error), error))
        .orDefault(defaultConfig)

    const configSubsection = readSubsection(config) // Config -> Either<unknown, Subsection>
}
```

This code can not be written as a single statement without `chainLeft`, because dropping out of the `Either` requires temporary variables spanning multiple statements. It is a contrived solution, so it does look silly, as one could break `getConfig` into its own method in this specific case.

Alternatively, with `chainLeft`:

```
function getConfigSubsection() {

    const config = readConfig()
        .toEither(`Could not read config file`)
        .chain(ConfigFile.decode)
        .mapLeft(error => (console.error(error), error))
        .chainLeft(() => defaultConfig)
        .chain(readSubsection)
}
```

which more accurately models intended data flow to the reader (again, in my opinion)

**Comment by gigobyte** [comment]
Alright, looks reasonable.

By the way (outside of the scope of this issue), I see that you're using the new Codec module in your example, are you currently using and what are your experiences so far?

**Comment by EricCrosson** [comment]
Woohoo! I'm glad to hear this has been accepted as a feature-request.

Good eye re the use of a Codec, I did use it for the first time this week.

Upon first read of the module's documentation (when 0.14 was released), I didn't really grok how it would help me. Then I found myself parsing a JSON string from a possibly-defined environment variable as a way to override default settings and realized the Codec is probably the perfect tool for the job. 

It took a little experimentation, but the current code is far more declarative and expressive than the imperative mess I started out with. I am grateful that the Codec definition gives me a logical place to put my parsing code, in a structure that can be hammered with tests and passed around. 10/10 will use again next time I have some code to parse from a program boundary. 


I'm glad you asked though because I do have a few questions I wasn't able to answer from the current state of the documentation....(time passes)....but that I was able to answer in the process of trying to form coherent questions.

My stumbling blocks consisted of my perception of the phrase `JSON value` in the Codec documentation. Perhaps because I was dealing with stringified JSON, I took the documentation to imply that a Codec's `decode` method (that "Takes a JSON value and runs the decode function") would be passed a JSON value in _stringified_ format, and then `decode` would run on the parsed JSON value. I was confused until just now why I had to manually call `JSON.parse`, why the `input` to `decode` was of type `unknown` and not `string`, and by the fact that `encode` for simple values didn't call `JSON.stringify(input)`.

I see now that JSON value doesn't mean serialized string notation of JSON value, but rather an object in valid JSON format. D'oh!

So now that Codec is making more sense to me than ever, I quite enjoy the addition. I haven't yet composed a codec from primitives (my use-case is just a NonEmptyList of strings of specific format, with no surrounding object) so I can't give feedback on that side, but the structure, functionality, and error messages are all well worth the learning curve.
